docs/architecture-decisions/2024-01-13_native_binaries.md

```diff
@@ -0,0 +1,183 @@
+- You don't need to install Python on your system to use AlgoKit if you aren't creating a Python project.
+
+## Options
+
+### Option 1 - PyInstaller
+
+**Pros**
+
+- Easy to use and configure
+- Supports multiple platforms and architectures
+- Can handle complex packages and dependencies
+- Generates a single file executable
+- Active development and community support
+- Fairly fast build time via ci - ~3-4 minutes
+- Fairly small executable size (see benchmarking results below)
+- Marginally equal executable load time in `onedir` mode compared to `onefile` mode
+
+**Cons**
+
+- Occasionally requires manual configuration for more complex packages
+- Requires complex build packaging matrix to support multiple platforms and architectures
+- Requires minor tweaks in algokit cli to account for the fact that features relying on `sys.executable` will point to algokit cli executable instead of python interpreter. This is a minor change and can be done in a backwards compatible way however still a con to consider.
+- Requires minor tweaks in algokit cli to introduce `multiprocessing.freeze_support()` to avoid issues with `vanity-address` task when executing via binary.
+
+#### PoC
+
+The PoC is available [here](https://github.com/algorandfoundation/algokit-cli/pull/382). It outlines a simple github action with extra setup that compiles algokit cli as a single file executable on latest versions of Windows, Mac and Linux github runners.
+
+### Option 2 - Nuitka
+
+**Pros**
+
+- Nuitka translates Python code into C and then compiles it, which can result in performance improvements.
+- Cross-Platform: Supports multiple platforms including Windows, macOS, and Linux.
+- More cross compilations options than PyInstaller
+- Official github action simplifies the process of building executables for different platforms.
+
+**Cons**
+
+- Compilation Time: The process of converting Python to C and then compiling can be time-consuming. Up to ~30 minutes on github with 3 parallel jobs.
+- Size of Executable: The resulting executables can be larger due to the inclusion of the Python interpreter and the compiled C code (see benchmarking results below).
+- Does not support Python 3.12.
+- Requires minor tweaks in algokit cli to account for the fact that features relying on `sys.executable` will point to algokit cli executable instead of python interpreter. This is a minor change and can be done in a backwards compatible way however still a con to consider.
+- Requires minor tweaks in algokit cli to introduce `multiprocessing.freeze_support()` to avoid issues with `vanity-address` task when executing via binary.
+
+#### PoC
+
+The PoC is available [here](https://github.com/algorandfoundation/algokit-cli/pull/393). It outlines a simple github action with extra setup that compiles algokit cli as a single file executable on latest versions of Windows, Mac and Linux github runners.
+
+### Option 3 - Cross platform python wheels with `cibuildwheel`
```

Comment:
Question for @achidlow @daniel-makerx , technically our pyproject.toml currently states that the build wheel is a pure python wheel. However we rely on dependencies that rely on non pure python based wheels like `cffi`, `cryptography`, `pycryptodomex` - however they all are distributed via manylinux build wheels and cover large amount of architectures/oses and etc. Does this imply that we can safely assume that algokit cli wheel is still pure python or reliance on packages as mentioned, implies we still need to build our own wheels in with tools like `cibuildwheel`? I can adjust this option to instead propose this tool https://github.com/cloud-custodian/poetry-plugin-freeze which we can use to simply freeze the wheel (does some tweaking with locking all specified dependencies prior to building the wheel) and we don't need to worry about manylinux stuff (assuming we are indeed building a pure python wheel)

Reply:
@achidlow i removed mentions of cibuildwheel as we clarified during the sync that algokit cli is indeed a pure python wheel